cpplint.py

```diff
@@ -934,9 +934,93 @@ def unicode_escape_decode(x):
 # This is set by --headers flag.
 _hpp_headers = set([])
 
-# {str, bool}: a map from error categories to booleans which indicate if the
-# category should be suppressed for every line.
-_global_error_suppressions = {}
+class ErrorSuppressions:
+  """Class to track all error suppressions for cpplint"""
+
+  class SuppressionRange:
+    """Class to represent a range of line numbers for which an error is suppressed"""
+    def __init__(self, begin=0, end=math.inf):
+      self._begin = begin
+      self._end = end
+
+    def __str__(self):
+      return f'[{self.begin}-{self.end}]'
+
+    def __contains__(self, obj):
+      if isinstance(obj, ErrorSuppressions.SuppressionRange):
+        return self.begin <= obj.begin and self.end >= obj.end
+      elif isinstance(obj, int):
+        return self.begin <= obj <= self.end
+      raise TypeError()
+
+    @property
+    def begin(self):
+      return self._begin
+
+    @property
+    def end(self):
+      return self._end
+
+    @end.setter
+    def end(self, end):
+      assert end > self.begin
+      self._end = end
+
+  def __init__(self):
+    self._suppressions = {}
+    self._open_block_suppressions = []
```

Comment:
why do we need to keep track of many open block suppressions, if only one block can be open at a time? (sorry I have not finished reading all the code, maybe later things would become clear). Currently it seem to me we don't need to track this at all, we can scan all suppressions any time to see if the last is open or not? The state of the runtime would become easier to reason about.

Reply:
I did it this way because it lined up with how single line ignores were being done. Also, in theory there could be some checks which are done outside of the line by line checks such as include what you use which could be eligible for suppression.

Reply:
I think there might have been some confusion in this comment thread. I think I originally thought the suggestion was to get rid of both `_suppressions` and `_open_block_suppressions` but now that I am rereading it I think the suggestion was just to remove `_open_block_suppressions`. The list of open_block_suppressions is not actually needed and can just be the current open range. I think we still want to keep of the current open range because it makes it easier on end of block to close all open ranges. Making that change now